File: akka-persistence-query/src/main/java/akka/persistence/query/javadsl/ReadJournal.java
/*
 * Copyright (C) 2009-2015 Typesafe Inc. <http://www.typesafe.com>
 */

package akka.persistence.query.javadsl;

import akka.persistence.query.Query;
import akka.persistence.query.Hint;
import akka.stream.javadsl.Source;
import scala.annotation.varargs;

/**
 * Java API
 * <p>
 * API for reading persistent events and information derived
 * from stored persistent events.
 * <p>
 * The purpose of the API is not to enforce compatibility between different
 * journal implementations, because the technical capabilities may be very different.
 * The interface is very open so that different journals may implement specific queries.
 * <p>
 * Usage:
 * <pre><code>
 * final ReadJournal journal =
 *   PersistenceQuery.get(system).getReadJournalFor(queryPluginConfigPath);
 *
 * final Source&lt;EventEnvelope, ?&gt; events =
 *   journal.query(new EventsByTag("mytag", 0L));
 * </code></pre>
 */

public interface ReadJournal {

  /**
   * Java API
   * <p>
   * A query that returns a `Source` with output type `T` and materialized value `M`.
   * <p>
   * The `hints` are optional parameters that defines how to execute the
   * query, typically specific to the journal implementation.
   *
   */
  <T, M> Source<T, M> query(Query<T, M> q, Hint... hints);

}

File: akka-persistence-query/src/main/scala/akka/persistence/query/javadsl/ReadJournal.scala
/*
 * Copyright (C) 2009-2017 Lightbend Inc. <http://www.lightbend.com>
 */
package akka.persistence.query.javadsl

/**
 * API for reading persistent events and information derived
 * from stored persistent events.
 *
 * The purpose of the API is not to enforce compatibility between different
 * journal implementations, because the technical capabilities may be very different.
 * The interface is very open so that different journals may implement specific queries.
 *
 * There are a few pre-defined queries that a query implementation may implement,
 * such as [[EventsByPersistenceIdQuery]], [[PersistenceIdsQuery]] and [[EventsByTagQuery]]
 * Implementation of these queries are optional and query (journal) plugins may define
 * their own specialized queries by implementing other methods.
 *
 * Usage:
 * {{{
 * SomeCoolReadJournal journal =
 *   PersistenceQuery.get(system).getReadJournalFor(SomeCoolReadJournal.class, queryPluginConfigPath);
 * Source<EventEnvolope, Unit> events = journal.eventsByTag("mytag", 0L);
 * }}}
 *
 * For Scala API see [[akka.persistence.query.scaladsl.ReadJournal]].
 */
trait ReadJournal

